File: DoublyGeneric/DoublyLinearGen.cpp
#include<iostream>
using namespace std;



template<class T>
struct node 
{
    T data;
    struct node<T> *next;
    struct node<T> *prev;
};

template<class T>
class DoublyLGen
{
    private :
        struct node<T> * first;
        int iCount;

    public:
        DoublyLGen();

        void InsertFirst(T);
        void InsertLast(T);
        void InsertAtPos(T, int);

        void DeleteFirst();
        void DeleteLast();
        void DeleteAtPos(int);

        void Display();
        int Count();
};


template<class T>
DoublyLGen<T> ::DoublyLGen()
{
    first = NULL;
    iCount =0;
}


template<class T>
void DoublyLGen<T> :: InsertFirst(T No)
{
    struct node<T> *newn = NULL;
    newn = new node<T>;
    newn->data = No;
    newn->next = NULL;
    newn->prev = NULL;

    if(first == NULL)
    {
        first = newn;
    }
    else
    {
        newn->next = first;
        first->prev = newn;
        first = newn;
    }
    iCount++;

}

template<class T>
void DoublyLGen<T> :: InsertLast(T No)
{

    struct node<T> *newn = NULL;
    newn = new node<T>;
    newn->data = No;
    newn->next = NULL;
    newn->prev = NULL;

    if(first == NULL)
    {
        first = newn;
    }
    else
    {
        struct node<T> *temp = first;
        while(temp ->next != NULL)
        {
            temp = temp ->next;
        }
        temp->next = newn;
        newn->prev = temp;
        
    }
    iCount++;

}

template<class T>
void DoublyLGen<T> :: InsertAtPos(T No, int Pos)
{
    if(Pos < 1 || Pos > iCount+1)
    {
        cout<<"Invalid Position....\n";
        return ;
    }

    if(Pos == 1)
    {
        InsertFirst(No);
    }
    else if(Pos == iCount+1)
    {
        InsertLast(No);
    }
    else    
    {
        struct node<T> *temp = first;
        struct node<T> *newn = NULL;
        newn = new node<T>;
        newn->data = No;
        newn->next = NULL;
        newn->prev = NULL;

        for(int iCnt = 1 ; iCnt < Pos -1 ; iCnt++)
        {
            temp = temp ->next;
        }
        newn->next = temp->next;
        newn->prev = temp;
        temp->next->prev = newn;
        temp->next = newn;
        iCount++;


    }
}


template<class T>
void DoublyLGen<T> :: DeleteFirst()
{
    if(first == NULL)
    {
        return;
    }
    if(first->next == NULL)
    {
        delete first;
        first = NULL;
    }
    else
    {
        struct node<T> *temp = first;

        first = first->next;
        first->prev = NULL;
        delete temp;
    }
    iCount--;
}


template<class T>
void DoublyLGen<T> :: DeleteLast()
{
    if(first == NULL)
    {
        return;
    }
    if(first->next == NULL)
    {
        delete first;
        first = NULL;
    }
    else
    {
        struct node<T> *temp = first;

        while(temp->next != NULL)
        {
            temp = temp ->next;
        }
        temp->prev->next = NULL;
        delete temp;
    }
    iCount--;
    
}


template<class T>
void DoublyLGen<T> :: DeleteAtPos(int Pos)
{
    if(Pos < 1 || Pos > iCount)
    {
        cout<<"Invalid Position....\n";
        return ;
    }

    if(Pos == 1)
    {
        DeleteFirst();
    }
    else if(Pos == iCount)
    {
        DeleteLast();
    }
    else    
    {
        struct node<T> *temp = first;
        struct node<T> *target = NULL;

        for(int iCnt = 1 ; iCnt < Pos -1 ; iCnt++)
        {
            temp = temp ->next;
        }
        target = temp->next;
        temp->next = target->next;
        if(target->next != NULL)
        {
            target->next->prev = temp;
        }

        delete target;
        iCount--;
    }
}


template<class T>
void DoublyLGen<T> :: Display()
{
    struct node<T> * temp = first;
    cout << "NULL <=> ";
    while (temp != NULL)
    {
        cout << "| " << temp->data << " | <=> ";
        temp = temp->next;
    }
    cout << "NULL\n";
}

template<class T>
int DoublyLGen<T> :: Count()
{
    return iCount;
}

int main()
{
    int datatype = 0;

    cout << "Select data type:\n";
    cout << "1. int\n";
    cout << "2. char\n";
    cout << "3. double\n";
    cout << "4. string\n";
    cout << "Enter choice: ";
    cin >> datatype;

    if (datatype == 1)
    {
        DoublyLGen<int> sobj;
        int choice = 1, value = 0, pos = 0;

        while (choice != 0)
        {
            cout << "\n-----------------------------------------------------------\n";
            cout << "\n----------------DOUBLY-LINEAR-IN-GENERIC-------------------\n";
            cout << "\n-----------------------------------------------------------\n";
            cout << "1. Insert First\n";
            cout << "2. Insert Last\n";
            cout << "3. Insert At Position\n";
            cout << "4. Delete First\n";
            cout << "5. Delete Last\n";
            cout << "6. Delete At Position\n";
            cout << "7. Display\n";
            cout << "8. Count\n";
            cout << "0. Exit\n";
            cout << "--------------------------------------\n";
            cout << "Enter your choice: ";
            cin >> choice;

            switch (choice)
            {
            case 1:
                cout << "Enter int: ";
                cin >> value;
                sobj.InsertFirst(value);
                break;
            case 2:
                cout << "Enter int: ";
                cin >> value;
                sobj.InsertLast(value);
                break;
            case 3:
                cout << "Enter int: ";
                cin >> value;
                cout << "Enter position: ";
                cin >> pos;
                sobj.InsertAtPos(value, pos);
                break;
            case 4:
                sobj.DeleteFirst();
                break;
            case 5:
                sobj.DeleteLast();
                break;
            case 6:
                cout << "Enter position: ";
                cin >> pos;
                sobj.DeleteAtPos(pos);
                break;
            case 7:
                sobj.Display();
                break;
            case 8:
                cout << "Count: " << sobj.Count() << endl;
                break;
            case 0:
                cout << "Thank you using Application....!\n";
                break;
            default:
                cout << "Invalid choice!\n";
                break;
            }
        }
    }
    else if (datatype == 2)
    {
        DoublyLGen<char> sobj;
        int choice = 1, pos = 0;
        char value;

        while (choice != 0)
        {
            cout << "\n-----------------------------------------------------------\n";
            cout << "\n----------------DOUBLY-LINEAR-IN-GENERIC-------------------\n";
            cout << "\n-----------------------------------------------------------\n";
            cout << "1. Insert First\n";
            cout << "2. Insert Last\n";
            cout << "3. Insert At Position\n";
            cout << "4. Delete First\n";
            cout << "5. Delete Last\n";
            cout << "6. Delete At Position\n";
            cout << "7. Display\n";
            cout << "8. Count\n";
            cout << "0. Exit\n";
            cout << "--------------------------------------\n";
            cout << "Enter your choice: ";
            cin >> choice;

            switch (choice)
            {
            case 1:
                cout << "Enter char: ";
                cin >> value;
                sobj.InsertFirst(value);
                break;
            case 2:
                cout << "Enter char: ";
                cin >> value;
                sobj.InsertLast(value);
                break;
            case 3:
                cout << "Enter char: ";
                cin >> value;
                cout << "Enter position: ";
                cin >> pos;
                sobj.InsertAtPos(value, pos);
                break;
            case 4:
                sobj.DeleteFirst();
                break;
            case 5:
                sobj.DeleteLast();
                break;
            case 6:
                cout << "Enter position: ";
                cin >> pos;
                sobj.DeleteAtPos(pos);
                break;
            case 7:
                sobj.Display();
                break;
            case 8:
                cout << "Count: " << sobj.Count() << endl;
                break;
            case 0:
                cout << "Thank you!\n";
                break;
            default:
                cout << "Invalid choice!\n";
                break;
            }
        }
    }
    else if (datatype == 3)
    {
        DoublyLGen<double> sobj;
        int choice = 1, pos = 0;
        double value;

        while (choice != 0)
        {
            cout << "\n-----------------------------------------------------------\n";
            cout << "\n----------------DOUBLY-LINEAR-IN-GENERIC-------------------\n";
            cout << "\n-----------------------------------------------------------\n";
            cout << "1. Insert First\n";
            cout << "2. Insert Last\n";
            cout << "3. Insert At Position\n";
            cout << "4. Delete First\n";
            cout << "5. Delete Last\n";
            cout << "6. Delete At Position\n";
            cout << "7. Display\n";
            cout << "8. Count\n";
            cout << "0. Exit\n";
            cout << "--------------------------------------\n";
            cout << "Enter your choice: ";
            cin >> choice;

            switch (choice)
            {
            case 1:
                cout << "Enter double: ";
                cin >> value;
                sobj.InsertFirst(value);
                break;
            case 2:
                cout << "Enter double: ";
                cin >> value;
                sobj.InsertLast(value);
                break;
            case 3:
                cout << "Enter double: ";
                cin >> value;
                cout << "Enter position: ";
                cin >> pos;
                sobj.InsertAtPos(value, pos);
                break;
            case 4:
                sobj.DeleteFirst();
                break;
            case 5:
                sobj.DeleteLast();
                break;
            case 6:
                cout << "Enter position: ";
                cin >> pos;
                sobj.DeleteAtPos(pos);
                break;
            case 7:
                sobj.Display();
                break;
            case 8:
                cout << "Count: " << sobj.Count() << endl;
                break;
            case 0:
                cout << "Thank you using Application....!\n";
                break;
            default:
                cout << "Invalid choice!\n";
                break;
            }
        }
    }
    else if (datatype == 4)
    {
        DoublyLGen<string> sobj;
        int choice = 1, pos = 0;
        string value;

        while (choice != 0)
        {
            cout << "\n-----------------------------------------------------------\n";
            cout << "\n----------------DOUBLY-LINEAR-IN-GENERIC-------------------\n";
            cout << "\n-----------------------------------------------------------\n";
            cout << "1. Insert First\n";
            cout << "2. Insert Last\n";
            cout << "3. Insert At Position\n";
            cout << "4. Delete First\n";
            cout << "5. Delete Last\n";
            cout << "6. Delete At Position\n";
            cout << "7. Display\n";
            cout << "8. Count\n";
            cout << "0. Exit\n";
            cout << "--------------------------------------\n";
            cout << "Enter your choice: ";
            cin >> choice;
            cin.ignore();

            switch (choice)
            {
            case 1:
                cout << "Enter string: ";
                getline(cin, value);
                sobj.InsertFirst(value);
                break;
            case 2:
                cout << "Enter string: ";
                getline(cin, value);
                sobj.InsertLast(value);
                break;
            case 3:
                cout << "Enter string: ";
                getline(cin, value);
                cout << "Enter position: ";
                cin >> pos;
                cin.ignore();
                sobj.InsertAtPos(value, pos);
                break;
            case 4:
                sobj.DeleteFirst();
                break;
            case 5:
                sobj.DeleteLast();
                break;
            case 6:
                cout << "Enter position: ";
                cin >> pos;
                cin.ignore();
                sobj.DeleteAtPos(pos);
                break;
            case 7:
                sobj.Display();
                break;
            case 8:
                cout << "Count: " << sobj.Count() << endl;
                break;
            case 0:
                cout << "Thank you using Application....!\n";
                break;
            default:
                cout << "Invalid choice!\n";
                break;
            }
        }
    }
    else
    {
        cout << "Invalid data type choice!\n";
    }

    return 0;
}

File: DoublyGeneric/DoublyLinearGen.java
import java.util.*;

class node<T>
{
    public T data;
    public node<T> next;
    public node<T> prev;
}

class DoublyLinear<T>
{
    private node<T> first;
    private int iCount;

    public DoublyLinear()
    {
        first = null;
        iCount = 0;
    }

    public void InsertFirst(T No)
    {
        node<T> newn = new node<T>();
        newn.data = No;

        if (first == null)
        {
            first = newn;
        }
        else
        {
            newn.next = first;
            first.prev = newn;
            first = newn;
        }
        iCount++;
    }

    public void InsertLast(T No)
    {
        node<T> newn = new node<T>();
        newn.data = No;

        if (first == null)
        {
            first = newn;
        }
        else
        {
            node<T> temp = first;

            while (temp.next != null)
            {
                temp = temp.next;
            }

            temp.next = newn;
            newn.prev = temp;
        }
        iCount++;
    }

    public void InsertAtPos(T No, int Pos)
    {
        if (Pos < 1 || Pos > iCount + 1)
        {
            System.out.println("Invalid Position...");
            return;
        }

        if (Pos == 1)
        {
            InsertFirst(No);
        }
        else if (Pos == iCount + 1)
        {
            InsertLast(No);
        }
        else
        {
            node<T> newn = new node<T>();
            newn.data = No;

            node<T> temp = first;

            for (int iCnt = 1; iCnt < Pos - 1; iCnt++)
            {
                temp = temp.next;
            }

            newn.next = temp.next;
            newn.prev = temp;
            temp.next.prev = newn;
            temp.next = newn;

            iCount++;
        }
    }

    public void DeleteFirst()
    {
        if (first == null)
        {
            return;
        }

        if (first.next == null)
        {
            first = null;
        }
        else
        {
            first = first.next;
            first.prev = null;
        }
        iCount--;
    }

    public void DeleteLast()
    {
        if (first == null)
        {
            return;
        }

        if (first.next == null)
        {
            first = null;
        }
        else
        {
            node<T> temp = first;

            while (temp.next != null)
            {
                temp = temp.next;
            }

            temp.prev.next = null;
        }
        iCount--;
    }

    public void DeleteAtPos(int Pos)
    {
        if (Pos < 1 || Pos > iCount)
        {
            System.out.println("Invalid Position...");
            return;
        }

        if (Pos == 1)
        {
            DeleteFirst();
        }
        else if (Pos == iCount)
        {
            DeleteLast();
        }
        else
        {
            node<T> temp = first;

            for (int iCnt = 1; iCnt < Pos - 1; iCnt++)
            {
                temp = temp.next;
            }

            node<T> target = temp.next;
            temp.next = target.next;
            target.next.prev = temp;

            iCount--;
        }
    }

    public void Display()
    {
        node<T> temp = first;

        System.out.print("NULL <=> ");

        while (temp != null)
        {
            System.out.print("| " + temp.data + " | <=> ");
            temp = temp.next;
        }

        System.out.println("NULL");
    }

    public int Count()
    {
        return iCount;
    }
}

public class DoublyLinearGen
{
    public static void main(String[] args)
    {
        Scanner sobj = new Scanner(System.in);

        System.out.println("Select data type:");
        System.out.println("1. Integer");
        System.out.println("2. Character");
        System.out.println("3. Double");
        System.out.println("4. String");
        System.out.print("Enter choice: ");
        int datatype = sobj.nextInt();
        sobj.nextLine(); // clear buffer

        int choice = 1;

        if (datatype == 1)
        {
            DoublyLinear<Integer> obj = new DoublyLinear<>();

            while (choice != 0)
            {
                showMenu();
                choice = sobj.nextInt();

                switch (choice)
                {
                    case 1:
                        System.out.print("Enter Integer: ");
                        obj.InsertFirst(sobj.nextInt());
                        break;

                    case 2:
                        System.out.print("Enter Integer: ");
                        obj.InsertLast(sobj.nextInt());
                        break;

                    case 3:
                        System.out.print("Enter Integer: ");
                        int val = sobj.nextInt();
                        System.out.print("Enter Position: ");
                        int pos = sobj.nextInt();
                        obj.InsertAtPos(val, pos);
                        break;

                    case 4:
                        obj.DeleteFirst();
                        break;

                    case 5:
                        obj.DeleteLast();
                        break;

                    case 6:
                        System.out.print("Enter Position: ");
                        int delPos = sobj.nextInt();
                        obj.DeleteAtPos(delPos);
                        break;

                    case 7:
                        obj.Display();
                        break;

                    case 8:
                        System.out.println("Count: " + obj.Count());
                        break;

                    case 0:
                        System.out.println("Thank you!");
                        break;

                    default:
                        System.out.println("Invalid choice!");
                        break;
                }
            }
        }
        else if (datatype == 2)
        {
            DoublyLinear<Character> obj = new DoublyLinear<>();

            while (choice != 0)
            {
                showMenu();
                choice = sobj.nextInt();

                switch (choice)
                {
                    case 1:
                        System.out.print("Enter Character: ");
                        obj.InsertFirst(sobj.next().charAt(0));
                        break;

                    case 2:
                        System.out.print("Enter Character: ");
                        obj.InsertLast(sobj.next().charAt(0));
                        break;

                    case 3:
                        System.out.print("Enter Character: ");
                        char ch = sobj.next().charAt(0);
                        System.out.print("Enter Position: ");
                        int pos = sobj.nextInt();
                        obj.InsertAtPos(ch, pos);
                        break;

                    case 4:
                        obj.DeleteFirst();
                        break;

                    case 5:
                        obj.DeleteLast();
                        break;

                    case 6:
                        System.out.print("Enter Position: ");
                        int delPos = sobj.nextInt();
                        obj.DeleteAtPos(delPos);
                        break;

                    case 7:
                        obj.Display();
                        break;

                    case 8:
                        System.out.println("Count: " + obj.Count());
                        break;

                    case 0:
                        System.out.println("Thank you!");
                        break;

                    default:
                        System.out.println("Invalid choice!");
                        break;
                }
            }
        }
        else if (datatype == 3)
        {
            DoublyLinear<Double> obj = new DoublyLinear<>();

            while (choice != 0)
            {
                showMenu();
                choice = sobj.nextInt();

                switch (choice)
                {
                    case 1:
                        System.out.print("Enter Double: ");
                        obj.InsertFirst(sobj.nextDouble());
                        break;

                    case 2:
                        System.out.print("Enter Double: ");
                        obj.InsertLast(sobj.nextDouble());
                        break;

                    case 3:
                        System.out.print("Enter Double: ");
                        double val = sobj.nextDouble();
                        System.out.print("Enter Position: ");
                        int pos = sobj.nextInt();
                        obj.InsertAtPos(val, pos);
                        break;

                    case 4:
                        obj.DeleteFirst();
                        break;

                    case 5:
                        obj.DeleteLast();
                        break;

                    case 6:
                        System.out.print("Enter Position: ");
                        int delPos = sobj.nextInt();
                        obj.DeleteAtPos(delPos);
                        break;

                    case 7:
                        obj.Display();
                        break;

                    case 8:
                        System.out.println("Count: " + obj.Count());
                        break;

                    case 0:
                        System.out.println("Thank you!");
                        break;

                    default:
                        System.out.println("Invalid choice!");
                        break;
                }
            }
        }
        else if (datatype == 4)
        {
            DoublyLinear<String> obj = new DoublyLinear<>();

            while (choice != 0)
            {
                showMenu();
                choice = sobj.nextInt();
                sobj.nextLine(); // clear buffer

                switch (choice)
                {
                    case 1:
                        System.out.print("Enter String: ");
                        obj.InsertFirst(sobj.nextLine());
                        break;

                    case 2:
                        System.out.print("Enter String: ");
                        obj.InsertLast(sobj.nextLine());
                        break;

                    case 3:
                        System.out.print("Enter String: ");
                        String str = sobj.nextLine();
                        System.out.print("Enter Position: ");
                        int pos = sobj.nextInt();
                        sobj.nextLine(); // clear buffer
                        obj.InsertAtPos(str, pos);
                        break;

                    case 4:
                        obj.DeleteFirst();
                        break;

                    case 5:
                        obj.DeleteLast();
                        break;

                    case 6:
                        System.out.print("Enter Position: ");
                        int delPos = sobj.nextInt();
                        sobj.nextLine(); // clear buffer
                        obj.DeleteAtPos(delPos);
                        break;

                    case 7:
                        obj.Display();
                        break;

                    case 8:
                        System.out.println("Count: " + obj.Count());
                        break;

                    case 0:
                        System.out.println("Thank you!");
                        break;

                    default:
                        System.out.println("Invalid choice!");
                        break;
                }
            }
        }
        else
        {
            System.out.println("Invalid data type!");
        }
    }

    public static void showMenu()
    {
        System.out.println("----------------------------------------------");
        System.out.println("------------ DOUBLY LINEAR GENERIC -----------");
        System.out.println("----------------------------------------------");
        System.out.println("1. Insert First");
        System.out.println("2. Insert Last");
        System.out.println("3. Insert At Position");
        System.out.println("4. Delete First");
        System.out.println("5. Delete Last");
        System.out.println("6. Delete At Position");
        System.out.println("7. Display");
        System.out.println("8. Count");
        System.out.println("0. Exit");
        System.out.print("Enter your choice: ");
    }
}
